### PR TITLE
Improve left/inner join performance by rerange right table by keys

### PR DESCRIPTION
**Pull request by KevinyhZou**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve the join performance by rerange the right table by keys while the table keys are dense in left or inner hash join.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Note about that we may have sparse columns and working with them through IColumn interface is critical.

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 7a5fc210ab92751bc90e39db6e119e5bbf3f23ea with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60341/7a5fc210ab92751bc90e39db6e119e5bbf3f23ea/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60341/7a5fc210ab92751bc90e39db6e119e5bbf3f23ea/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/RowRefs.h [hunk 1]
```diff
@@ -112,6 +113,17 @@ struct RowRefList : RowRef
...
         }
 
+        Batch * getBatch()
+        {
+            return batch;
+        }
+
+        void nextBatch()
+        {
+            batch = batch->next;
```
Comment (jkartseva) [inline src/Interpreters/RowRefs.h:123]: This invalidates the existing `position, using `operator ++` and `nextBatch()` together is not viable.

**Review by jkartseva — COMMENTED** [review]
Some initial comments.

File: src/Interpreters/RowRefs.h [hunk 2]
```diff
@@ -55,14 +56,14 @@ struct RowRefList : RowRef
 
         Batch * insert(RowRef && row_ref, Arena & pool)
         {
-            if (full())
+            if (full() || (size > 0 && row_ref.block != row_refs[0].block))
```
Comment (jkartseva) [inline src/Interpreters/RowRefs.h:59]: The new condition should be in the form a function with a human-readable name.

**Comment by KevinyhZou** [comment]
we have optimize inner join use batch insert  base on the pr https://github.com/ClickHouse/ClickHouse/pull/58278, and on our gluten case , it can performance better. please review it . @jkartseva

**Comment by jkartseva** [comment]
Is the most recent iteration ready for review @KevinyhZou ?

**Comment by KevinyhZou** [comment]
> Is the most recent iteration ready for review @KevinyhZou ?
yes @jkartseva

**Comment by vdimir** [comment]
I assume some CI failures can be related, for example

> Stateful tests (tsan) — Tests are not finished, fail: 1, passed: 7, skipped: 1 [Details](https://s3.amazonaws.com/clickhouse-test-reports/60341/f266807e0cc64ad8ddcff741594bcc85508bd6ab/stateful_tests__tsan_.html)

Contains thread sanitizer report around `HashJoin::joinBlock`, see `ThreadSanitizer: data race` in stderr.log file:
https://s3.amazonaws.com/clickhouse-test-reports/60341/f266807e0cc64ad8ddcff741594bcc85508bd6ab/stateful_tests__tsan_/stderr.log

Cloud you please check it?

**Comment by KevinyhZou** [comment]
ok

**Comment by KevinyhZou** [comment]
OK

**Comment by KevinyhZou** [comment]
cc @vdimir

**Comment by KevinyhZou** [comment]
the performance improves in `hashjoin_with_large_output`
image

the performance improves in `all_join_opt`
image

base on pr: https://github.com/ClickHouse/ClickHouse/pull/63677,   could you help review this pr, @jkartseva

**Comment by KevinyhZou** [comment]
The right table rerange cost a lot of time, but the join output rows of it is very small proportion compare to its total rows. So the rerange would not make impact on performance improve, but cause the performance degradation. I think this is the reason for `q13` performance degradation.

**Comment by KevinyhZou** [comment]
I had add a setting `join_to_sort_table_rows_threshold` to control not to rerange the right table if the right table has too many rows greater than this threshold, so we only rerange it when the table not very big, and not to make performance  degradation. and I have test on the sqls you mentioned, now not performance degradation.  @XiaoqiMa

**Comment by KevinyhZou** [comment]
any comments about this pr ? @jkartseva

**Comment by jkartseva** [comment]
I'll take a look by the end of the week @KevinyhZou

**Review by jkartseva — COMMENTED** [review]
I think we should make this feature experimental. I can help with rolling it out to the staging tier in our cloud. If there are no regressions, we may deprecate the experimental flag.

File: src/Core/Settings.h [hunk 3]
```diff
@@ -922,6 +922,8 @@ class IColumn;
     M(Bool, implicit_transaction, false, "If enabled and not already inside a transaction, wraps the query inside a full transaction (begin + commit or rollback)", 0) \
     M(UInt64, grace_hash_join_initial_buckets, 1, "Initial number of grace hash join buckets", 0) \
     M(UInt64, grace_hash_join_max_buckets, 1024, "Limit on the number of grace hash join buckets", 0) \
+    M(Int32, join_to_sort_perkey_rows_threshold, 40, "The lower limit of per-key average rows in the right table to determine whether to sort it in hash join.", 0) \
+    M(Int32, join_to_sort_table_rows_threshold, 10000, "The upper limit of rows in the right table to determine whether to sort it in hash join.", 0) \
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Core/Settings.h:926]: The meaning of the thresholds is unclear without reading the code. We could consider one of the following options: updating the description to clarify how changing the setting affects user-experience (for example, using a special join method that improves performance for wide tables but increases memory consumption) or, even better, removing the thresholds and choosing the best value automatically.

[the remark this answers is not in the document]
Comment (KevinyhZou) [reply src/Core/Settings.h]: I tested locally and found that if there are many rows on the right but very few matching rows, sorting will lead to performance degradation. In this scenario, I found through testing that 10000 of `join_to_sort_table_rows_threshold` is a reasonable value and will not cause significant performance degradation due to sorting. On the contrary, if there are lots of matching rows to output, the threshold can be increased to allow a larger right table to be sorted, which can still achieve significant performance improvement.

And another threshold was set default value 40, as I test on the table when the table is not dense enough, and the sorting may also cause the performance degradation, when the threshold set up to 40, then no performance slow down.   @jkartseva

[the remark this answers is not in the document]
Comment (KevinyhZou) [reply src/Core/Settings.h]: I have update the description of the `threshold` settings, take a look at it whether would be ok? @vdimir

**Comment by KevinyhZou** [comment]
cc @jkartseva

File: src/Core/Settings.h [hunk 4]
```diff
@@ -922,6 +922,9 @@ class IColumn;
     M(Bool, implicit_transaction, false, "If enabled and not already inside a transaction, wraps the query inside a full transaction (begin + commit or rollback)", 0) \
     M(UInt64, grace_hash_join_initial_buckets, 1, "Initial number of grace hash join buckets", 0) \
     M(UInt64, grace_hash_join_max_buckets, 1024, "Limit on the number of grace hash join buckets", 0) \
+    M(Int32, join_to_sort_perkey_rows_threshold, 40, "Rerange the right table by key in left or inner hash join when the per-key average rows of it exceed this value (means the table keys is dense) and its number of rows is not too many(controlled by `join_to_sort_table_rows_threshold`), to make the join output by the data batch of key, which would improve performance.", 0) \
+    M(Int32, join_to_sort_table_rows_threshold, 10000, "Rerange the right table by key in left or inner hash join when its number of rows not exceed this value and the table keys is dense (controlled by `join_to_sort_perkey_rows_threshold`), to make the join performance improve as output by the data batch of key, but not cost too much on the table reranging.", 0) \
```
Comment (jkartseva) [inline src/Core/Settings.h:926]: Similarly:
_The upper threshold of the number of rows in the right table to determine whether to rerange the right table by key in left or inner join._

Or:
_The maximum number of rows in the right table..._

"upper" or "max" should be in the setting name.

File: src/Core/SettingsChangesHistory.cpp [hunk 5]
```diff
@@ -94,7 +94,10 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
...
+            {"join_to_sort_perkey_rows_threshold", 0, 40, "Rerange the right table by key in left or inner hash join when the per-key average rows of it exceed this value (means the table keys is dense) and its number of rows is not too many(controlled by `join_to_sort_table_rows_threshold`), to make the join output by the data batch of key, which would improve performance."},
+            {"join_to_sort_table_rows_threshold", 0, 10000, "Rerange the right table by key in left or inner hash join when its number of rows not exceed this value and the table keys is dense (controlled by `join_to_sort_perkey_rows_threshold`), to make the join performance improve as output by the data batch of key, but not cost too much on the table reranging."},
+            {"allow_experimental_join_right_table_sorting", false, false, "If it is set to true, and the conditions of `join_to_sort_perkey_rows_threshold` and `join_to_sort_perkey_rows_threshold` are met, then we will try to rerange the right table by key to improve the performance in left or inner hash join."},
```
Comment (jkartseva) [inline src/Core/SettingsChangesHistory.cpp:100]: Let's remove "try" from the description:

_"...are met, rerange the right table by key..."_

**Review by jkartseva — COMMENTED** [review]
I think this looks good, let's adjust the setting names (see comment) and provide cleaner descriptions, and I'll approve & merge.

**Review by jkartseva — APPROVED** [review]
Looks good, thank you for working on this.

**Comment by jkartseva** [comment]
Please update the Changelog entry with a more generalized summary. The present description is too verbose and focused on the specific case.

**Comment by KevinyhZou** [comment]
done

File: src/Core/Settings.h [hunk 6]
```diff
@@ -922,6 +922,9 @@ class IColumn;
     M(Bool, implicit_transaction, false, "If enabled and not already inside a transaction, wraps the query inside a full transaction (begin + commit or rollback)", 0) \
     M(UInt64, grace_hash_join_initial_buckets, 1, "Initial number of grace hash join buckets", 0) \
     M(UInt64, grace_hash_join_max_buckets, 1024, "Limit on the number of grace hash join buckets", 0) \
+    M(Int32, join_to_sort_minimum_perkey_rows, 40, "The lower limit of per-key average rows in the right table to determine whether to rerange the right table by key in left or inner join. This setting ensures that the optimization is not applied for sparse table keys", 0) \
```
Comment (Algunenano) [inline src/Core/Settings.h:925]: Same as https://github.com/ClickHouse/ClickHouse/pull/63677/files#r1771844385. Why is this a Int32 it it's treated as unsigned and it shouldn't be negative?

Comment (KevinyhZou) [reply src/Core/Settings.h:925]: Yes, the type is wrong. And I have made a pr to change the type, https://github.com/ClickHouse/ClickHouse/pull/69886

**Pull request merged**